internal/impl/gcp/input_spanner_change_stream.go

```diff
@@ -0,0 +1,277 @@
+		// there is still the chance that we could lose changes though.
+		changeChannel: make(chan *model.DataChangeRecord, 1),
+		log:           log,
+		shutdownSig:   shutdown.NewSignaller(),
+	}
+	if out.partitionDSN, err = conf.FieldString("partition_dsn"); err != nil {
+		return
+	}
+
+	if out.partitionTable, err = conf.FieldString("partition_table"); err != nil {
+		return
+	}
+
+	if out.streamDSN, err = conf.FieldString("stream_dsn"); err != nil {
+		return
+	}
+
+	if out.streamID, err = conf.FieldString("stream_id"); err != nil {
+		return
+	}
+
+	if out.allowedModTypes, err = conf.FieldStringList("allowed_mod_types"); err != nil {
+		return
+	}
+
+	useInMemPartition, err := conf.FieldBool("use_in_mememory_partition")
+	if err != nil {
+		return
+	}
+
+	startTimeEpoch, err := conf.FieldInt("start_time_epoch")
+	if err != nil {
+		return
+	}
+
+	if startTimeEpoch > 0 {
+		out.startTime = func(seconds int) *time.Time {
+			t := time.UnixMicro(int64(startTimeEpoch))
+			return &t
+		}(startTimeEpoch)
+	}
+
+	if !useInMemPartition && slices.Contains([]string{out.partitionDSN, out.partitionTable, out.streamDSN, out.streamID}, "") {
+		return nil, errors.New("partition_dsn, partition_table, stream_dsn, and stream_id must be set")
+	} else if slices.Contains([]string{out.streamDSN, out.streamID}, "") {
+		return nil, errors.New("stream_dsn, and stream_id must be set")
+	}
+	out.usePartitionTable = !useInMemPartition
+	return
+}
+
+func init() {
+	err := service.RegisterInput(
+		"gcp_spanner_change_stream", newSpannerChangeStreamInputConfig(),
+		func(conf *service.ParsedConfig, mgr *service.Resources) (service.Input, error) {
+			return newSpannerStreamInput(conf, mgr.Logger())
+		})
+	if err != nil {
+		panic(err)
+	}
+}
+
+//------------------------------------------------------------------------------
+
+type spannerStreamInput struct {
+	streamDSN         string
+	streamClient      *db
+	streamID          string
+	partitionDSN      string
+	partitionTable    string
+	usePartitionTable bool
+	startTime         *time.Time
+	allowedModTypes   []string
+	reader            streamReader
+	// create a channel to pass from connection to read.
+	changeChannel chan *model.DataChangeRecord
+
+	log         *service.Logger
+	shutdownSig *shutdown.Signaller
+}
+
+func (i *spannerStreamInput) Connect(ctx context.Context) (err error) {
+	jobctx, _ := i.shutdownSig.SoftStopCtx(context.Background())
+
+	if i.streamClient == nil {
+		i.streamClient, err = newDatabase(jobctx, i.streamDSN)
+		if err != nil {
+			return err
+		}
+	}
+	if i.reader == nil {
+		i.reader, err = newStreamer(jobctx, i.streamClient, i.streamID, i.partitionDSN, i.partitionTable, i.usePartitionTable, i.allowedModTypes, i.startTime)
+		if err != nil {
+			return err
+		}
+	}
+	go func() {
+		if rerr := i.reader.Stream(jobctx, i.changeChannel); rerr != nil {
+			i.log.Errorf("Subscription error: %v\n", rerr)
+			close(i.changeChannel)
+			panic(rerr)
```

Comment:
We don't panic in Connect, please gracefully mark the input as disconnected, then return `service.ErrNotConnected` in `Read` so that the input is restarted.

Reply:
If you want to see an example of doing this - see the postgres_cdc input.